### PR TITLE
Implement robust cheat searching interface with rumble features

### PR DESCRIPTION
**Pull request by RetroSven**
## Guidelines

1. Rebase before opening a pull request
2. If you are sending several unrelated fixes or features, use a branch and a separate pull request for each
3. If possible try squashing everything in a single commit. This is particularly beneficial in the case of feature merges since it allows easy bisecting when a problem arises

## Description

Here is the list of new features with this commit:

 - Cleaned up a couple of files to remove some compilation warnings

 - "Load Cheats" has been broken into two separate options.  One will replace the list of cheats with what is being loaded whereas the other will append to the list of cheats with what is being loaded.

 - Cheat list is now automatically saved and loaded for each game

 - When using left/right to change the values of numbers, there is now a feature that gradually increases the step-size the longer you hold the button

 - Full cheat list management - Edit, Delete, Copy, Add, (to top of list, to bottom of list, above entry, below entry)

 - There are now two different cheat types - Emulator (handled by the emulator, the original type) and the new type - RetroArch - that is handled by RetroArch itself

 - Robust cheat searching interface (see instructions below)

 - Rumble activation based on memory changing.  For example, after finding the memory location for the number of lives in a game (via the cheat searching interface) you can set it up such that every time the value decreases (lose a life) the controller rumbles.  Rumble tested with X360 controller, input driver dinput, joypad driver xinput.  Available rumble controls:
	Rumble when memory value changes
	Rumble when memory value does not change
	Rumble when memory value decreases
	Rumble when memory value increases
	Rumble when memory value = value
	Rumble when memory value != value 
	Rumble when memory value < value
	Rumble when memory value > value

General Guideline for Finding New Cheats:

1) Start game

2) Go to Quick Menu -> Cheats -> Start or Continue Cheat Search 

3) Use left/right on "Start or Restart Cheat Search" to select a bit-size appropriate to the console you are using and the value your are searching.  
For example, if you are playing Castlevania:SOTN on the PS1 and you want to search for the health value, then that's a value that can be greater than 255 (0xFF), but it's unlikely that the game developers anticipated a value larger than 65535 (0xFFFF) so set the search to 16-bit.
An alternate example - if you are playing Space Invaders on Atari 2600 and you want to search for the number of lives, then that's a value that could possibly be stored in just 2-bits of data (max number of lives = 3) and since the Atari 2600 only has a very small memory space, it's entirely possible that the memory location for the number of lives is only partially stored in a single byte while the rest of that same byte may store other important data that should not be touched.  Set the size to 2-bit.

4) Select "Start or Restart Cheat Search" once you have selected the bit size

5) Go back to the game and lose a life

6) Go back to the quick menu and select "Search Memory For Values ... Less Than Before" because when you started the search you had one more life than you do now.  You could also try "Search Memory For Values ... Equals To Before-1".  The number of matches should go down.

7) If the number of matches is still too great to peruse, then perform actions 5 and 6 repeatedly until the number of matches is something you feel comfortable trying (e.g. 10).  If you run out of lives, just reset the game or restore a save state.  Then your lives will likely be greater than the last time you checked, so select "Search Memory For Values ... Greater Than Before"

8) Once you have a manageable list, select "Add the ## Matches to Your List"

9) Go back one menu to see the codes that have been added.  Try turning just one on at a time to see if it has the desired effect.  If not, turn it off and try the next one.  One of them should be the location in memory that stores your number of lives and enabling the cheat in its default state will result in that memory location being overwritten by the cheat value continuously and voila infinite lives.

Alternately, you can "Search Memory For Values ... Equal to ###" if you know the exact number (e.g. the number of hit points you have in an RPG). 

## Related Issues

n/a

## Related Pull Requests

n/a

## Reviewers

not sure

### CODE REVIEW

**Comment by twinaphex** [comment]
Looks cool. Could you do me a favor though and fix the resolve conflict? Seems it's something in intl/msg_hash_us.h.

@fr500 What do you think about this PR? Any feedback, thoughts?

**Comment by i30817** [comment]
What's the point of using rumble? I mean most cheats finding depend on the user narrowing down the possibilities with the help of information from the game itself, the cheat interface notifying the user that the 'value changed' is not useful if you're still narrowing down the value since the user has to deduce that the value is present on memory and to manipulate it ingame.

**Comment by RetroSven** [comment]
Sorry wasn't sure if the conflict was my responsibility - didn't want to do something that wasn't SOP.

WRT rumble - it's not used for the cheat-code searching - it's used for game enhancement.  For example, once you have found the memory location where your health is stored in an ARPG, then every time an enemy causes you damage you can make the controller rumble a little.  There's a wide area of application.  Back when I was doing this sort of thing on the original xbox, people found it very entertaining.
Thanks

**Comment by i30817** [comment]
i added a paragraph to my comment, i'd be glad if you commented on that, specifically if all those conditions :
Rumble when memory value changes
Rumble when memory value does not change
Rumble when memory value decreases
Rumble when memory value increases
Rumble when memory value = value
Rumble when memory value != value
Rumble when memory value < value
Rumble when memory value > value

couldn't be collapsed to 'rumble when the cheat system prevents a write from happening' (ie: the value is set to a fixed value) or 'rumble when the value changes by any means (normal gameplay as in the example you gave)'.

Rumble when 'nothing changed' doesn't seem particularly useful to me but it could be a third setting...

**Comment by twinaphex** [comment]
I think it is actually a novel and neat idea.

**Comment by RetroSven** [comment]
Note that both the cheat and rumble functionality can be explicitly disabled.  Neither is reliant nor necessarily related to the other except in that they both reference the same location in the game's memory.  I.e. you can have a cheat that pins a value to a specific part of memory without any rumble functionality and you could also have a rumble that does not have any memory-altering functionality that the cheat action performs.

That being said, all of the different changes to memory are necessary in order to target specific actions in memory.  You might want to know when the value increases for slight rumble when you drink a potion.  Decrease in health.  Perhaps there is a value from 0-255 where any value from 1-255 means that you're invincible and 0 means not - you'd want a "not equal to 0" option so that you could rumble when you're invincible (think of "star power" in super mario)  There's so much that can be done with this but all of the fine-grained controls really are necessary.

Thanks

**Comment by RetroSven** [comment]
As an aside, people loved this feature back when I implemented it in some xbox projects from long ago.  :P

I think it will go over well in the retro community.  :)

File: setting_list.c [hunk 1]
```diff
@@ -173,11 +183,85 @@ void setting_get_string_representation_size_in_mb(void *data,
...
+            step = setting->step*10000.0f ;
+         } else if ( global->menu.action_press_time  > 15000000)
+         {
+            step = setting->step*1000.0f ;
+         } else if ( global->menu.action_press_time  > 12000000)
+         {
+            step = setting->step*100.0f ;
+         } else if ( global->menu.action_press_time  > 9000000)
+         {
+            step = setting->step*100.0f ;
```
Comment (RobLoach) [inline setting_list.c]: Looks like the 12000000 and 9000000 conditions do the same thing?

In addition, they seem like a lot of [magic numbers](https://en.wikipedia.org/wiki/Magic_number_%28programming%29). Could there be a way to clean up their use? If not, that's understandable.

**Comment by twinaphex** [comment]
> As an aside, people loved this feature back when I implemented it in some xbox projects from long ago. 📦 

Oh, which xbox projects were those? I really loved the Xbox OG homebrew scene.

**Comment by RetroSven** [comment]
Remember XPort?

"twinaphex" sounds familiar - did you frequent the forums on xbox-scene.com back when it was alive?

**Comment by saftle** [comment]
Will this work on the rumble codes that were made for Znexbox for the original Xbox OG? Here are some rumble codes I found: 
[rumble.zip](https://github.com/libretro/RetroArch/files/2235020/rumble.zip)

**Comment by RetroSven** [comment]
I didn't do that one (didn't touch SNES until the very end out of respect for Lantus), but after some experimentation it looks like yes they use the same memory addresses.  For example, in Castlevania X the first one is 168.  Entering that as the memory address in RA does prevent your health/energy from going down.  Here's an AWK script which I believe accurately decodes the files:

```
#rtype
#0 = when memory changes
#1 = when memory decreases
#2 = when memory increases
{
   if ( NR %10 == 4 ) desc=$0;
   if ( NR %10 == 2 ) address=$1;
   if ( NR %10 == 3 ) rtype=$1;
   if ( NR %10 == 5 ) strength1=$1;
   if ( NR %10 == 6 ) strength2=$1;
   if ( NR %10 == 7 ) duration1=$1;
   if ( NR %10 == 8 ) duration2=$1;

   if ( NR %10 == 0 ){
      print desc  " " rtype " " address ;
   }
}
```
strings -n 1 "Castlevania - Dracula X (U).rmb" | gawk -f /path/to/the/above/awk/file.awk

**Comment by RetroSven** [comment]
Oh and I also found a bug while investigating those  zsnexbox rumble codes - if you start RA and open a pre-existing cheat, then try to edit the memory address, it's range-locked at min=0, max=0 (because at that point the cheat system has not been initialized so it doesn't know what the max address is).  This is fixed now, but I'm holding off on commits so that all changes/fixes to this PR can be done in one commit.

**Comment by saftle** [comment]
Awesome. I absolutely loved the rumble features in Znexbox and the like. Thank you so much for contributing to the xbox og scene and now to RetroArch @RetroSven !

**Comment by RetroSven** [comment]
Thanks saftle  :)

**Comment by twinaphex** [comment]
@RetroSven Yeah I was only an enduser back then.

Anyway, great to have you here. I merged your PR.

**Comment by twinaphex** [comment]
@RetroSven had to push some C89 buildfixes -

https://github.com/libretro/RetroArch/commit/89083fd093c4343e4b4a473dbdda7a0b1373c66f

Let me know if everything still works fine on your end as far as this PR is concerned in master now.

**Comment by RetroSven** [comment]
re:  XPort definitely did a lot of great stuff back then.

(btw I meant Xport=me  :P  )

I certainly had much more free time back then to do those projects.  When the Nintendo Switch homebrew started to  take off  I began thinking about converting some  of my old xbox stuff since it seems to  be a nice portable unit for retrogaming.  While looking into it I learned about  Lakka, RetroArch and the nx port.  A lot has happened in the emu scene  since I left!  I figured no sense reinventing the wheel, thus these recent contributions.

Also happy to find out that there is a Dreamcast emu that fixed that bug with Record of  Lodoss War  that made it unplayable.  Wii emularors - lots of good stuff.

Anyway yes things still  seem to work the same after those C89 fixes.  I've made a  mental  note about the styling ( brackets and such)

about to  make another  PR to fix a bug

Thanks for a great retro platform!

**Comment by fr500** [comment]
Hey

It's creating empty cheat files in the game directory.
This is totally unacceptable. There is a cheat directory defined for a reason.

**Comment by daninthemix** [comment]
Great that someone's doing work on the cheat system. I have one tiny request: can we get a 'reload cheats' or 'reload file' option? Often when I'm setting up a game, I have to edit the cheat file multiple, and it would be great to not always have to load cheats...locate the file, turn the options all back on, etc.

**Comment by RetroSven** [comment]
@fr500 I was actually using the path that the game defined to be the cheat file (global->name.cheatfile) and since the cheat directory to which you're referring is named "settings->paths.path_cheat_database" I wasn't sure if I should be using it for things that aren't from the actual online cheat database.  I've changed it so that it saves to [path_cheat_database]/[core name]/[game name].cht - will make a PR in a bit.

@daninthemix - all the cheat properties can be managed via the UI - I'm unclear why you need to hand-edit and then reload multiple times when you can just edit whatever you're changing via the UI.  All of your cheats are saved on a game-specific manner now and are auto-saved upon exit and auto-loaded upon loading.  So basically if you're setting up a new game, you'd pull the game from the cheat database, enable/delete/change whatever you need from that and then when you exit all of those cheat settings are auto-saved to "gamename.cht" so that the next time you load the game they are all restored and ready to go.

**Comment by daninthemix** [comment]
Because editing / typing / copying and pasting a cheat file is so, so much quicker in Notepad++ than doing anything on a control pad, or inside a UI. It can often take a lot of this to get a working cheat set, as there's a lot of non-working codes out there etc.

But it's certainly not a biggie at all.

**Comment by fr500** [comment]
Thanks for addressing that.

**Comment by RetroSven** [comment]
@twinaphex  - do you have an opinion about what @daninthemix has suggested?

**Comment by twinaphex** [comment]
OK, I'd say if it is easy to add, let's add it to the UI then. Is that something you are able to add by yourself?

**Comment by RetroSven** [comment]
yup will add

**Comment by daninthemix** [comment]
One further request: auto-apply codes whenever we toggle them. Particularly now the 'apply' button is buried in the middle of the menu. Some games really do demand that you toggle certain codes on and off quite often.

**Comment by daninthemix** [comment]
Thankyou!

**Comment by RetroSven** [comment]
@saftle @twinaphex what repo houses the cheats in the cheat database?  I've converted those SNES rumble codes to RA format and wanted to submit them

**Comment by daninthemix** [comment]
I've found a problem with the new changes. Look at how this cheat file has been truncated after loading, compared to stable 1.7.3 (it isn't applying the final 2 codes).
[retroarch-cheats-issue.zip](https://github.com/libretro/RetroArch/files/2259210/retroarch-cheats-issue.zip)

**Comment by RetroSven** [comment]
@daninthemix fixed in 
https://github.com/libretro/RetroArch/pull/7041

**Comment by daninthemix** [comment]
Great. Does it now allow cheats of the same length as before? Because some cheats on gamehacking.org (which allows you to download codes in libretro cheat format) are very, very long so there is a need for long codes (or many-parted codes), hopefully without having to faff around and break them into sub-codes.

Couple more things:

1 - cheats don't auto-apply until I visit the cheats menu and click 'apply changes' - even though it has remembered which cheats were active last time I played the game, and is listing them as 'enabled'. They aren't active till I click apply.

2 - for the 'reload cheat file' option I actually meant the user-selected file, rather than the cheats\core\game auto-generated one. Doesn't matter though.

**Comment by RetroSven** [comment]
@daninthemix yes any sized code - there are some in Pokemon Pearl that are over 9000 characters long - crazy

also auto-applying during load option:
https://github.com/libretro/RetroArch/pull/7053

**Comment by daninthemix** [comment]
Really awesome work :)

I can't grab your latest changes as I'm using a development build with the VRR runloop and don't want to lose that.

@twinaphex any chance of [that](https://github.com/libretro/RetroArch/pull/7042) getting merged? Also are you aware of the certificate errors on buildbot? Chrome won't even allow you to bypass and go to the site for dev builds.

**Comment by twinaphex** [comment]
I am asking @bparker06 and @Tatsuya79 if it can be merged, I believe @bparker06 still had to test something before it could be merged, so I just need to know the status of that.

**Comment by daninthemix** [comment]
@RetroSven I have another suggestion: for long codes, it can take ages for it to scroll over and show you whether its [ON] or [OFF]. We need to see at a glance when scrolling down which codes are enabled.

How about simply making the active codes a different color - yellow, for example?

**Comment by RetroSven** [comment]
@daninthemix 
https://github.com/libretro/RetroArch/pull/7060
Seemed simpler to just reverse the display - enabled state is first, then the code

**Comment by RetroSven** [comment]
@twinaphex 
Every libretro project should now have memory hooks in order to use this new cheat finder and rumble functionality (except for some older MAME and fbalpha branches)

If I missed any or there are problems with any please let me know

**Comment by daninthemix** [comment]
@RetroSven finally got a chance to test this and the cheat status isn't displaying at all now (see attached

).

**Comment by RetroSven** [comment]
oops:
https://github.com/libretro/RetroArch/pull/7074

**Comment by daninthemix** [comment]
Really great work. Have tested the latest build with Beetle and Mupen and cheats work great, and you've added some really valuable quality-of-life improvements.

One more suggestion - actually it's more a discussion point, whether you agree this would be an improvement - I think RetroArch should enumerate the cheats by itself, rather than have the user sequentially number them, and write in the total at the top of the cheats file, as this is quite annoying when building / iterating on a cheat file.

But I also suspect this may involve quite a lot of code changes so I'll just leave it with you.

Thanks again!

**Comment by twinaphex** [comment]
Great that you saw this through @RetroSven ! Would it be possible for you to write a quick write-up for us  (in terms of how to use it) so that we can feature this in our upcoming RetroArch 1.7.4 blogpost? Thanks already in advance!

**Comment by RetroSven** [comment]
@daninthemix thanks - however I'm not clear on your suggestion - can you walk me through the steps or your use case scenario?

@twinaphex yup, I'll amend the description in this PR to fill in the changes since it was originally made

**Comment by RetroSven** [comment]
@twinaphex :
RetroArch now has built-in cheat code searching/creating.  Additional cheat system updates:



- "Load Cheats" has been broken into two separate options. One will replace the list of cheats with what is being loaded whereas the other will append to the list of cheats with what is being loaded.

- Cheat list is now automatically saved and loaded for each game

- When using left/right to change the values of numbers, there is now a feature that gradually increases the step-size the longer you hold the button

- Full cheat list management - Edit, Delete, Copy, Add, (to top of list, to bottom of list, above entry, below entry)

- There are now two different cheat types - Emulator (handled by the emulator, the original type) and the new type - RetroArch - that is handled by RetroArch itself

- Option to reload cheat file (if you manually edit file while RA is running)

- Option to auto-apply cheats when loading game

- Option to auto-apply cheats when toggling them on/off

- Robust cheat searching interface (see instructions below) which works for just about every RA core

- Rumble activation based on memory changing. For example, after finding the memory location for the number of lives in a game (via the cheat searching interface) you can set it up such that every time the value decreases (lose a life) the controller rumbles. Rumble tested with X360 controller, input driver dinput, joypad driver xinput. Available rumble controls:
 Rumble when memory value changes
    Rumble when memory value does not change
    Rumble when memory value decreases
    Rumble when memory value increases
    Rumble when memory value = value
    Rumble when memory value != value
    Rumble when memory value < value
    Rumble when memory value > value

- General Guideline for Finding New Cheats:

    Start game

    Go to Quick Menu -> Cheats -> Start or Continue Cheat Search

    Use left/right on "Start or Restart Cheat Search" to select a bit-size appropriate to the console you are using and the value your are searching.
    For example, if you are playing Castlevania:SOTN on the PS1 and you want to search for the health value, then that's a value that can be greater than 255 (0xFF), but it's unlikely that the game developers anticipated a value larger than 65535 (0xFFFF) so set the search to 16-bit.
    An alternate example - if you are playing Space Invaders on Atari 2600 and you want to search for the number of lives, then that's a value that could possibly be stored in just 2-bits of data (max number of lives = 3) and since the Atari 2600 only has a very small memory space, it's entirely possible that the memory location for the number of lives is only partially stored in a single byte while the rest of that same byte may store other important data that should not be touched. Set the size to 2-bit.

    Select "Start or Restart Cheat Search" once you have selected the bit size

    Go back to the game and lose a life

    Go back to the quick menu and select "Search Memory For Values ... Less Than Before" because when you started the search you had one more life than you do now. You could also try "Search Memory For Values ... Equals To Before-1". The number of matches should go down.

    If the number of matches is still too great to peruse, then perform actions 5 and 6 repeatedly until the number of matches is something you feel comfortable trying (e.g. 10). If you run out of lives, just reset the game or restore a save state. Then your lives will likely be greater than the last time you checked, so select "Search Memory For Values ... Greater Than Before"

    Once you have a manageable list, select "Add the ## Matches to Your List"

    Go back one menu to see the codes that have been added. Try turning just one on at a time to see if it has the desired effect. If not, turn it off and try the next one. One of them should be the location in memory that stores your number of lives and enabling the cheat in its default state will result in that memory location being overwritten by the cheat value continuously and voila infinite lives.

    Alternately, you can "Search Memory For Values ... Equal to ###" if you know the exact number (e.g. the number of hit points you have in an RPG).

- Over 1000 new cheat and rumble codes added to the database across the following systems:
Atari 5200
Atari 7800
Atari 400/800/1200XL
ColecoVision
MSX/MSX2
NES
SNES
Sega Master System
Sega Genesis
TurboGrafx-16/PC-Engine

Many thanks to madmab, Bombbloke, Gilou9999, Bigby, Megaman_? for providing cheat and rumble codes for several different systems!

**Comment by daninthemix** [comment]
@RetroSven Currently, you need to load cheats files that look like this:

```
cheats = 9

cheat0_desc = "Infinite Health"
cheat0_enable = true
cheat0_code = "800CC49A 0006"

cheat1_desc = "Infinite Oxygen"
cheat1_enable = true
cheat1_code = "800CC382 0001"

etc
```

I'm suggesting RetroArch should also be able to handle this:

```
cheat_desc = "Infinite Health"
cheat_enable = true
cheat_code = "800CC49A 0006"

cheat_desc = "Infinite Oxygen"
cheat_enable = true
cheat_code = "800CC382 0001"
```

No total needed, and RetroArch enumerates the cheats by itself, rather than requiring the user to increment them.

**Comment by RetroSven** [comment]
OK understood.  Something like that is of course do-able, but as you suspected would take a sizeable re-write and isn't the ideal method of parsing the general RA config file format (which is essentially what these cheat files are).  I agree that something is needed to make it easier to just copy/paste or add cheats to these cheat files outside of RA itself.  Perhaps it would mean having a separate loading/saving system just for the cheat file formats, but I'll noodle over it.
Thanks

**Comment by daninthemix** [comment]
@RetroSven I know this is beyond the scope of what you're doing, but if you got a chance to look at the Parallel core that would be great - currently it only applies the first cheat in the list, and I very much doubt anyone else will fix this.

**Comment by RetroSven** [comment]
@daninthemix I perused the code (did not run) and it nothing jumps out at me - are you sure the problem isn't with the codes themselves?  Do you have a use case to reproduce?

**Comment by daninthemix** [comment]
Actually no - I can't reproduce. It's one of those things that stuck in my mind because multiple times when using this core, I've loaded a cheat file and had weird stuff happen - either it crashed when I enabled the second code, or the second code simply didn't work (as was the case this morning, after which I posted the above).

That was Bangai-O, I had 2 codes, and only one worked when I enabled it. Just now, when I tried again - obviously this time the codes auto-loaded - I was able to toggle it on and it worked just fine. So perhaps there's an issue only when selecting a cheat file (on this core), or perhaps there's no issue at all.

Sorry!

**Pull request merged**